lib/sycamore/sycamore/transforms/detr_partitioner.py

```diff
@@ -168,6 +244,276 @@ def partition_pdf(
+        batch_size: int = 1,
+    ) -> List[List["Element"]]:
+        if not table_structure_extractor:
+            table_structure_extractor = DEFAULT_TABLE_STRUCTURE_EXTRACTOR(device=self.device)
+
+        pdfminer = None
+        exec = ProcessPoolExecutor(max_workers=1)
+        if not use_ocr:
+            with LogTime("start_pdfminer", log_start=True):
+                pdfminer = exec.submit(self._run_pdfminer, filename)
+
+        deformable_layout = []
+        if tracemalloc.is_tracing():
+            before = tracemalloc.take_snapshot()
+        for i in self._batch_pdf_convert(filename, batch_size):
+            parts = self.process_batch(
+                i,
+                threshold,
+                use_ocr,
+                ocr_images,
+                ocr_tables,
+                extract_table_structure,
+                table_structure_extractor,
+                extract_images,
+            )
+            assert len(parts) == len(i)
+            deformable_layout.extend(parts)
+            if tracemalloc.is_tracing():
+                gc.collect()
+                after = tracemalloc.take_snapshot()
+                top_stats = after.compare_to(before, "lineno")
+
+                print("[ Top 10 differences ]")
+                for stat in top_stats[:10]:
+                    print(stat)
+                before = after
+                display_top(after)
+
+        if pdfminer is not None:
+            with LogTime("wait_for_pdfminer", log_start=True):
+                pdfminer_layout = pdfminer.result()
+            assert len(pdfminer_layout) == len(deformable_layout), f"{len(pdfminer_layout)} vs {len(deformable_layout)}"
+            with LogTime("pdfminer_supplement"):
+                for d, p in zip(deformable_layout, pdfminer_layout):
+                    self._supplement_text(d, p)
+
+        if tracemalloc.is_tracing():
+            (current, peak) = tracemalloc.get_traced_memory()
+            logging.info(f"Memory Usage current={current} peak={peak}")
+            top = tracemalloc.take_snapshot()
+            display_top(top)
+        return deformable_layout
+
+    def _batch_pdf_convert(self, filename: str, batch_size: int) -> Generator[List[Image.Image], None, None]:
+        """Note: model service will call this to get batches of images for processing"""
+        batch = []
+        for i in self._convert_from_path_streamed(filename):
+            batch.append(i)
+            if len(batch) == batch_size:
+                yield batch
+                batch = []
+
+        if len(batch) > 0:
+            yield batch
+
+    def _convert_from_path_streamed(self, pdf_path: str) -> Generator[Image.Image, None, None]:
+        class StdoutEOF:
+            pass
+
+        class StderrEOF:
+            pass
+
+        def capture_exception(q, fn, finish_msg):
+            try:
+                fn()
+            except e:
+                q.put(e)
+
+            q.put(finish_msg)
+
+        def read_stdout(fh, q):
```

Comment:
same as above, do we need all this in the detr_partitioner file? It's making it hard to read just the partitioning logic

Reply:
Moved to utils.pdf